File: solutions/807/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Dec 12 11:36:08 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 
#include <algorithm> 
#include <numeric> 

using namespace std; 

class Solution {
public:
    int maxIncreaseKeepingSkyline(vector<vector<int>>& grid) {
        vector<int> rowMax(grid.size(), 0);
        vector<int> colMax(grid[0].size(), 0);
        for (int i = 0; i < grid.size(); ++i)
            rowMax[i] = *max_element(grid[i].begin(), grid[i].end());
        for (int i = 0; i < grid[0].size(); ++i) {
            int maxSum = INT_MIN;
            for (int j = 0; j < grid.size(); ++j) {
                if (maxSum < grid[j][i]) maxSum = grid[j][i];
            }
            colMax[i] = maxSum;
        }

        int result = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                result += min(rowMax[i], colMax[j]) - grid[i][j];
            }
        }

        return result;
    }
};

TEST(Test807, SimpleTest) {
    vector<vector<int>> grid = {{3,0,8,4},{2,4,5,7},{9,2,6,3},{0,3,1,0}};
    const int ans = 35; 
    Solution s; 
    EXPECT_EQ(s.maxIncreaseKeepingSkyline(grid), ans);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/807/main.py
from typing import List 

class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        rowMax = [0]*len(grid)
        colMax = [0]*len(grid[0])

        for i in range(len(grid)):
            rowMax[i] = max(grid[i])
        for i in range(len(grid[0])):
            colMax[i] = max([grid[j][i] for j in range(len(grid))])

        total = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                total += min(rowMax[i], colMax[j]) - grid[i][j]
                grid[i][j] = min(rowMax[i], colMax[j])

        return total

if __name__ == "__main__": 
    grid = [[3,0,8,4],[2,4,5,7],[9,2,6,3],[0,3,1,0]]
    s = Solution()
    ans = 35
    assert s.maxIncreaseKeepingSkyline(grid) == ans 

